text/0000-trust-boundary-specification.md

```diff
@@ -0,0 +1,128 @@
+
+### The new trust boundary
+
+The new trust boundary will be separated into multiple levels, with increasing permissiveness. Each level includes the permissions of the previous levels.
+
+- At the first level, a package is allowed to do operations that can have undefined results.
+- At the second level, a package is allowed to do operations that can have undefined behaviour.
+- At the third level, a package is allowed to do FFI calls.
+
+The main reason for separating undefined behaviour operations and the FFI into different levels is for general side-effect control. An FFI call can do _anything_, including writing to files or to the network, but a Pony operation with potential undefined behaviour has a limited scope.
+
+Out-of-memory errors aren't covered by the trust boundary because trying to do so would greatly limit the usefulness of untrusted packages.
+
+## User-facing changes
+
+### In the compiler
+
+With the trust boundary now being split into multiple levels, the `--safe` flag must also be separated into multiple levels. The new flags, `--safe-1`, `--safe-2`, and `--safe-3`, can be used in the same way as `--safe` to specify the trust level of packages in the program. `--safe` will be removed.
+
+- Higher-level flags take precedence over lower-level flags.
+- In the same way `--safe` works currently, if any of those flags is specified, any package not in the trusted list is treated as untrusted.
+- If none of those flags are specified, all packages are treated as trusted with a level of 3.
+- The `builtin` package is always treated as trusted with a level of 3.
+
+### In the language
+
+New syntax is needed to mark unsafe operations. We propose to use new annotations for this purpose: `unsafe-1`, `unsafe-2` and `unsafe-3`. These annotations map directly to the various levels of the trust boundary. These annotations are only allowed on function definitions, for example `fun \unsafe-1\ foo()`.
+
+Flags (`safe`) and annotations (`unsafe`) use antonyms because the intended wording of this syntax is "the most trusted package is allowed to use the most unsafe operation".
+
+- An unsafe function can only be called from a package with the corresponding trust level or higher.
+- A package can only define unsafe functions if it has the corresponding trust level or higher.
+- Safety level is taken into account in function subtyping, with safer functions (e.g. `unsafe-1`) being subtypes of less safe functions (e.g. `unsafe-3`).
+
+`unsafe-3` is intended for operations that can have an unlimited scope. For example, an operation composed of multiple FFI calls.
+
+### In the standard library
+
+Unsafe maths functions on numeric primitives will be marked `unsafe-1`. There is nothing requiring to be marked `unsafe-2` or `unsafe-3` currently so these annotations will stay unused for now.
```

Comment:
> There is nothing requiring to be marked `unsafe-2` or `unsafe-3` currently

This doesn't seem quite right. We have standard library types that do FFI, and standard library types that do pointer math.

Am I missing something?

Reply:
The idea is to only mark functions that expose the unsafety to the user (i.e. that have a narrow contract). `Array` does raw pointer operations, but these operations are fully controlled by the preliminary checks that `Array` enforces and as a result can't cause problems (i.e. `Array` has a wide contract).

This is a generalisation of the checks for FFI. You can only use the FFI from a safe package, but this isn't transitive: a package being safe means that you trust that package to use the FFI correctly and to expose a safe interface to other packages. The same thing would apply to any unsafe operation.

Reply:
I think this answer does more to answer my other comment than this one.

If we ignore pointer math for the moment and focus on FFI, can you explain why you don't think the FFI-using functions in the `net` package need `unsafe-3`?

Reply:
Because I can't cause my program to crash or to have some other kind of UB by using these functions since the inputs to the FFI calls are fully controlled. In addition, since you need to have access to `AmbientAuth` (or something derived from it) in order to use the stuff in `net`, there is no need for another layer of protection through `--safe`.

Reply:
How can the compiler verify that those inputs are controlled properly?

It is my understanding of this RFC that the compiler is going to enforce that all functions that call FFI include the `unsafe-3` annotation. Is that wrong? It sounds now like you're saying that adding the annotation is optional, based on whether you as the code author think you created a safe wrapper for the function.

If that's how this is intended to work, I think this is a step backward from how the `--safe` flag works now, since it enforces that no packages use FFI other than the ones you whitelist. If the new trust boundary works based on these annotations, and the annotations are optional, then you're not able to enforce anything, and an uncareful or unscrupulous package author can sneak an unsafe Pony API through a `--safe-{x}`-protected compilation simply by leaving out the annotations...

Reply:
The compiler will enforce that

- Every package that calls `unsafe-1` functions is marked `--safe-1` or higher
- Every package that calls `unsafe-2` functions is marked `--safe-2` or higher
- Every package that calls `unsafe-3` functions or uses the FFI is marked `--safe-3` or higher

The compiler won't be able to enforce that the exposed interface is safe, but that doesn't change anything for the FFI. In the current state of the compiler and language, I can have a function that calls `sqrt` without any input validation, and even if I mark the package it is declared in as trusted, it won't prevent untrusted packages from using it with negative numbers and causing UB. This RFC will not change anything on that front.

Reply:
Okay, thanks for spelling it out further. I understand what's being proposed now.

The part I didn't understand is the point that FFI calls would be treated as `unsafe-3` function calls, without needing to be annotated.